compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -1372,39 +1377,20 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         otherwise_block: BasicBlock,
         candidates: &mut [&mut Candidate<'_, 'tcx>],
     ) {
-        // If any candidate starts with an or-pattern, we have to expand the or-pattern before we
-        // can proceed further.
-        let expand_ors = candidates.iter().any(|candidate| {
-            matches!(
-                &*candidate.match_pairs,
-                [MatchPair { test_case: TestCase::Or { .. }, .. }, ..]
-            )
-        });
         ensure_sufficient_stack(|| {
-            if !expand_ors {
-                // No candidates start with an or-pattern, we can continue.
-                self.match_expanded_candidates(
-                    span,
-                    scrutinee_span,
-                    start_block,
-                    otherwise_block,
-                    candidates,
-                );
-            } else {
-                self.expand_and_match_or_candidates(
-                    span,
-                    scrutinee_span,
-                    start_block,
-                    otherwise_block,
-                    candidates,
-                );
-            }
+            self.match_candidates_with_enough_stack(
+                span,
+                scrutinee_span,
+                start_block,
+                otherwise_block,
+                candidates,
+            )
         });
     }
 
-    /// Construct the decision tree for `candidates`. Caller must ensure that no candidate in
-    /// `candidates` starts with an or-pattern.
-    fn match_expanded_candidates(
+    /// Construct the decision tree for `candidates`. Don't call this, call `match_candidates`
+    /// instead to reserve sufficient stack space.
+    fn match_candidates_with_enough_stack(
```

Comment:
Nitpick: I find this name a bit confusing, because it's not immediately obvious whether it *ensures* sufficient stack, or *assumes* sufficient stack (correct answer), or *filters* for sufficient stack.

Suggestion: I would normally call this something like `match_candidates_inner` or `match_candidates_impl` or `do_match_candidates`.

Reply:
Fair point, I was hoping the comment would be clear enough